### PR TITLE
Adds concurrency to hydration requests

### PR DESCRIPTION
**Pull request by another-rex**
An attempt at concurrency for hydration requests. I'm not sure if this is the most concise way to do so, it seems pretty long for what should be a simple concurrency logic.

### CODE REVIEW

File: pkg/osv/osv.go [hunk 1]
```diff
@@ -185,23 +187,74 @@ func Get(id string) (*models.Vulnerability, error) {
...
 
+	responseChan := make(chan hydrateResponse, maxConcurrentRequests)
+	rateLimiter := make(chan struct{}, maxConcurrentRequests)
+
+	// Run in separate goroutine so that we can immediately start
+	// listening to the responseChan, and it will be not blocked
+	go func(resp *BatchedResponse) {
+		for batchIdx, response := range resp.Results {
+			for resultIdx, vuln := range response.Vulns {
+				rateLimiter <- struct{}{}
```
Comment (oliverchang) [inline pkg/osv/osv.go:210]: nit: can we use https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool instead?

File: pkg/osv/osv.go [hunk 2]
```diff
@@ -185,23 +187,74 @@ func Get(id string) (*models.Vulnerability, error) {
...
+				go func(id string, batchIdx int, resultIdx int) {
+					vuln, err := Get(id)
+					if err != nil {
+						responseChan <- hydrateResponse{
```
Comment (oliverchang) [inline pkg/osv/osv.go:214]: I don't think we need this channel. 

Can we just pre-allocate the hydrated structure (i.e. all indexes are accessible) and have each "worker" write to the specified index directly? 

I believe this is safe, per the example in https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool:

```golang
	// Compute the output using up to maxWorkers goroutines at a time.
	for i := range out {
		// When maxWorkers goroutines are in flight, Acquire blocks until one of the
		// workers finishes.
		if err := sem.Acquire(ctx, 1); err != nil {
			log.Printf("Failed to acquire semaphore: %v", err)
			break
		}

		go func(i int) {
			defer sem.Release(1)
			out[i] = collatzSteps(i + 1)
		}(i)
	}
```

And this: https://go.dev/ref/spec#Variables:

> Structured variables of [array](https://go.dev/ref/spec#Array_types), [slice](https://go.dev/ref/spec#Slice_types), and [struct](https://go.dev/ref/spec#Struct_types) types have elements and fields that may be [addressed](https://go.dev/ref/spec#Address_operators) individually. Each such element acts like a variable.

And all the nested structures in `HydratedBatchedResponse` are slices.

**Review by oliverchang — COMMENTED** [review]
nice! is this in response to any particular performance complaints in particular?

**Comment by another-rex** [comment]
Not any issue in particular, I just realized the sequential calls are a major part of the run time when looking at implementing #303 .

File: pkg/osv/osv.go (hunk 1 above)
Comment (another-rex) [reply pkg/osv/osv.go:210]: I did see that package, but since semaphores are under the /x/ experimental package I'm not sure whether we should use it.

Comment (oliverchang) [reply pkg/osv/osv.go:210]: Should be fair game to use it. There's a lot of prominent packages using it already: https://deps.dev/go/golang.org%2Fx%2Fsync/v0.1.0/dependents

File: pkg/osv/osv.go (hunk 2 above)
Comment (another-rex) [reply pkg/osv/osv.go]: Good idea, done

**Review by oliverchang — APPROVED** [review]

**Comment by oliverchang** [comment]
@another-rex are we ready to merge?

**Review by G-Rath — APPROVED** [review]

**Pull request merged**